e2store/src/era2.rs

```diff
@@ -0,0 +1,313 @@
+use std::{
+    io::{Read, Write},
+    path::PathBuf,
+};
+
+use alloy_primitives::{hex, B256, U256};
+use alloy_rlp::{Decodable, RlpDecodable, RlpEncodable};
+use anyhow::ensure;
+use core::panic;
+use ethportal_api::types::{execution::header::Header, state_trie::account_state::AccountState};
+
+use crate::{
+    e2store::{
+        stream::E2StoreStream,
+        types::{Entry, VersionEntry},
+    },
+    era1::HeaderEntry,
+};
+
+// <network-name>-<block-number>-<short-state-root>.era2
+//
+// era2 := Version | CompressedHeader | account*
+// account :=  CompressedAccount | CompressedStorage*
+// -----
+// Version                     = { type: 0x3265, data: nil }
+// CompressedHeader            = { type: 0x03,   data: snappyFramed(rlp(header)) }
+// CompressedAccount           = { type: 0x08,   data: snappyFramed(rlp(address_hash, rlp(nonce,
+// balance, storage_root, code_hash), raw_bytecode, storage_entry_count) }
+// CompressedStorage           = { type: 0x09,   data: snappyFramed(rlp(Vec<StorageItem {
+// storage_index_hash, value }>)) }
+// -----
+// CompressedStorage can have a max of 10k storage items, records must be filled before
+// creating a new one
+
+/// Represents an era2 `Era2` state snapshot.
+/// Unlike era1, not all fields will be stored in the struct, account's will be streamed from an
+/// iterator as needed.
+pub struct Era2 {
+    pub version: VersionEntry,
+    pub header: HeaderEntry,
+
+    /// e2store_stream, manages the interactions between the era2 state snapshot
+    e2store_stream: E2StoreStream,
+    storage_entries_left_to_decode: u16,
```

Comment:
nit: The `storage_entries_left_to_decode` doesn't make sense when we are writing to era2 file (because of the **decode**).
Maybe just rename it to `pending_storage_entries` or something like that.